jsk_perception/src/concave_hull_mask_image.cpp

```diff
@@ -0,0 +1,160 @@
+ *     disclaimer in the documentation and/o2r other materials provided
+ *     with the distribution.
+ *   * Neither the name of the JSK Lab nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+#include "jsk_perception/concave_hull_mask_image.h"
+#include <boost/assign.hpp>
+#include <boost/tuple/tuple.hpp>
+#include <jsk_topic_tools/log_utils.h>
+#include <jsk_recognition_utils/cv_utils.h>
+#include <opencv2/opencv.hpp>
+#include <sensor_msgs/image_encodings.h>
+#include <cv_bridge/cv_bridge.h>
+
+namespace jsk_perception
+{
+  void ConcaveHullMaskImage::onInit()
+  {
+    DiagnosticNodelet::onInit();
+
+    srv_ = boost::make_shared <dynamic_reconfigure::Server<Config> > (*pnh_);
+    dynamic_reconfigure::Server<Config>::CallbackType f =
+        boost::bind (&ConcaveHullMaskImage::configCallback, this, _1, _2);
+    srv_->setCallback(f);
+
+    pub_ = advertise<sensor_msgs::Image>(*pnh_, "output", 1);
+    onInitPostProcess();
+  }
+
+  void ConcaveHullMaskImage::configCallback(Config& config, uint32_t level)
+  {
+    boost::mutex::scoped_lock lock(mutex_);
+    min_area_ = config.min_area;
+    max_area_ = config.max_area;
+    dilate_size_ = config.dilate_size;
+  }
+
+  void ConcaveHullMaskImage::subscribe()
+  {
+    sub_ = pnh_->subscribe("input", 1, &ConcaveHullMaskImage::rectify, this);
+    ros::V_string names = boost::assign::list_of("~input");
+    jsk_topic_tools::warnNoRemap(names);
+  }
+
+  void ConcaveHullMaskImage::unsubscribe()
+  {
+    sub_.shutdown();
+  }
+
+  void ConcaveHullMaskImage::rectify(
+    const sensor_msgs::Image::ConstPtr& mask_msg)
+  {
+    vital_checker_->poke();
+    cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(
+      mask_msg, sensor_msgs::image_encodings::MONO8);
+    cv::Mat mask = cv_ptr->image;
+    cv::Mat dilate_img;
+
+    // Dilate to fill holes
+    cv::dilate(mask, dilate_img, getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(dilate_size_, dilate_size_)));
```

Comment:
Yeah, but it also can be done by `DilateMaskImage` node, so there must be special reason if you put here.